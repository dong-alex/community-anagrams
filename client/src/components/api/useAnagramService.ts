import ConfiguredAxios from "./ConfiguredAxios";
import { TopAnagram, Anagram } from "../../types";

// handles all api requests
const useAnagramService = () => {
  const getAllAnagrams = async (): Promise<Anagram[] | never> => {
    try {
      const { data }: { data: Anagram[] } = await ConfiguredAxios.get("/");
      console.log(data);
      return data || [];
    } catch (err) {
      console.log("Error in 'getAllAnagrams': ", err);
      throw Error(err.message);
    }
  };

  const saveAnagram = async (
    firstWord: string,
    secondWord: string
  ): Promise<boolean | never> => {
    try {
      const { data }: { data: boolean } = await ConfiguredAxios.post(
        "/anagrams",
        {
          firstWord,
          secondWord,
        }
      );
      return data;
    } catch (err) {
      console.log("Error in 'saveAnagram': ", err);
      throw Error(err.message);
    }
  };

  const getTopAnagrams = async (): Promise<TopAnagram[] | never> => {
    try {
      const { data }: { data: TopAnagram[] } = await ConfiguredAxios.get(
        "/top"
      );
      console.log(data);
      return data || [];
    } catch (err) {
      console.log("Error in 'getTopAnagrams': ", err);
      throw Error(err.message);
    }
  };

  return {
    getAllAnagrams,
    saveAnagram,
    getTopAnagrams,
  };
};

export default useAnagramService;
