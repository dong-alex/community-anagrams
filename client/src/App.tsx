import React, { FunctionComponent, useEffect, useState } from "react";
import { MDBBox, MDBContainer } from "mdbreact";
import Header from "./components/Header";
import TopAnagramsBoard from "./components/TopAnagramsBoard";
import AnagramRequestCard from "./components/AnagramRequestCard";
import useAnagramService from "./components/api/useAnagramService";
import isAlphabetical from "./components/utils/isAlphabetical";
import { TopAnagram } from "./types";

const App: FunctionComponent = () => {
  const { getTopAnagrams, saveAnagram } = useAnagramService();
  const [loading, setLoading] = useState<boolean>(false);
  const [topAnagrams, setTopAnagrams] = useState<TopAnagram[]>([]);

  useEffect(() => {
    const fetch = async () => {
      const data = await getTopAnagrams();
      setTopAnagrams(data);
      setLoading(false);
    };

    fetch();
  }, []);

  const handleSaveAnagram = async (
    firstWord: string,
    secondWord: string
  ): Promise<boolean | never> => {
    if (!isAlphabetical(firstWord) || !isAlphabetical(secondWord)) {
      throw Error(
        "Please remove any numbers or special symbols from your words and try again."
      );
    }

    try {
      setLoading(true);

      const result = await saveAnagram(firstWord, secondWord);

      setLoading(false);
      return result;
    } catch (err) {
      throw err;
    }
  };

  return (
    <MDBBox>
      <Header />
      <MDBContainer>
        <AnagramRequestCard onSaveAnagram={handleSaveAnagram} />
        <TopAnagramsBoard topAnagrams={topAnagrams} />
      </MDBContainer>
    </MDBBox>
  );
};

export default App;
