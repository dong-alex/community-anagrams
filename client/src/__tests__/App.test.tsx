import React from "react";
import { render, screen, act, fireEvent } from "@testing-library/react";
import App from "../App";
import useAnagramService from "../components/api/useAnagramService";

jest.mock("../components/api/useAnagramService");

describe("Overall App Tests", () => {
  beforeEach(() => {
    (useAnagramService as jest.Mock).mockRestore();
  });

  it("renders the app without data", async () => {
    (useAnagramService as jest.Mock).mockReturnValue({
      getAllAnagrams: jest.fn().mockResolvedValue([]),
      getTopAnagrams: jest.fn().mockResolvedValue([]),
      saveAnagram: jest.fn().mockResolvedValue(false),
    });

    await act(async () => {
      const { container } = render(<App />);
      expect(container).toMatchSnapshot();
    });
  });

  it("lists no top anagram requests initially", async () => {
    (useAnagramService as jest.Mock).mockReturnValue({
      getAllAnagrams: jest.fn().mockResolvedValue([]),
      getTopAnagrams: jest.fn().mockResolvedValue([]),
      saveAnagram: jest.fn().mockResolvedValue(false),
    });

    await act(async () => {
      render(<App />);
    });

    expect(screen.queryAllByTestId("top-anagram-item")).toHaveLength(0);
    expect(screen.getByTestId("top-anagrams-board-header")).toMatchSnapshot();
  });

  it("lists top anagram requests when they exist", async () => {
    const stub = jest.fn().mockResolvedValue([
      { firstWord: "testing", secondWord: "testing", count: 10 },
      { firstWord: "tester", secondWord: "tester", count: 9 },
    ]);

    (useAnagramService as jest.Mock).mockReturnValue({
      getAllAnagrams: jest.fn().mockResolvedValue([]),
      getTopAnagrams: stub,
      saveAnagram: jest.fn().mockResolvedValue(false),
    });

    await act(async () => {
      render(<App />);
    });

    expect(stub).toBeCalledTimes(1);
    const items = await screen.findAllByTestId("top-anagram-item");
    expect(items).toMatchSnapshot();
  });

  it("should handle saving anagrams when valid", async () => {
    (useAnagramService as jest.Mock).mockReturnValue({
      getAllAnagrams: jest.fn().mockResolvedValue([]),
      getTopAnagrams: jest.fn().mockResolvedValue([]),
      saveAnagram: jest.fn().mockResolvedValue(false),
    });

    render(<App />);
    const firstInputField = screen.getByTestId("first-word-input");
    const secondInputField = screen.getByTestId("second-word-input");
    const submitButton = screen.getByTestId("anagram-submission-button");

    fireEvent.change(firstInputField, {
      target: { value: "racer", name: "firstWord" },
    });

    fireEvent.change(secondInputField, {
      target: { value: "carer", name: "secondWord" },
    });

    await act(async () => {
      fireEvent.click(submitButton);
    });

    expect(screen.getByTestId("result-alert")).toMatchSnapshot();
  });

  it("should handle not saving anagrams when invalid", async () => {
    const spy = jest.fn();

    (useAnagramService as jest.Mock).mockReturnValue({
      getAllAnagrams: jest.fn().mockResolvedValue([]),
      getTopAnagrams: jest.fn().mockResolvedValue([]),
      saveAnagram: spy,
    });

    render(<App />);
    const firstInputField = screen.getByTestId("first-word-input");
    const secondInputField = screen.getByTestId("second-word-input");
    const submitButton = screen.getByTestId("anagram-submission-button");

    fireEvent.change(firstInputField, {
      target: { value: "ra2cer", name: "firstWord" },
    });

    fireEvent.change(secondInputField, {
      target: { value: "carer", name: "secondWord" },
    });

    await act(async () => {
      fireEvent.click(submitButton);
    });

    expect(spy).not.toHaveBeenCalled();
    expect(screen.getByTestId("error-alert")).toMatchSnapshot();
  });

  it("should handle error when saving anagram", async () => {
    // TODO: change all test descriptions to use "should"
    const stub = jest.fn().mockRejectedValue("error when submitting");
    (useAnagramService as jest.Mock).mockReturnValue({
      getAllAnagrams: jest.fn().mockResolvedValue([]),
      getTopAnagrams: jest.fn().mockResolvedValue([]),
      saveAnagram: stub,
    });

    render(<App />);
    const firstInputField = screen.getByTestId("first-word-input");
    const secondInputField = screen.getByTestId("second-word-input");
    const submitButton = screen.getByTestId("anagram-submission-button");

    fireEvent.change(firstInputField, {
      target: { value: "racer", name: "firstWord" },
    });

    fireEvent.change(secondInputField, {
      target: { value: "carer", name: "secondWord" },
    });

    await act(async () => {
      fireEvent.click(submitButton);
    });

    expect(stub).toHaveBeenCalledTimes(1);
    expect(screen.getByTestId("error-alert")).toMatchSnapshot();
  });
});
