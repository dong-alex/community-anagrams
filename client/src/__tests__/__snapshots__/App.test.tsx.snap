// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overall App Tests lists no top anagram requests initially 1`] = `
<h5
  class="text-center pt-5 pb-5"
  data-testid="top-anagrams-board-header"
>
  Top 10 Most Requested Anagram Checks
</h5>
`;

exports[`Overall App Tests lists top anagram requests when they exist 1`] = `
Array [
  <li
    class="list-group-item d-flex justify-content-between list-group-item-undefined"
    data-test="list-group-item"
    data-testid="top-anagram-item"
    style="user-select: none;"
  >
    <span
      class="ml-3"
    >
      1
    </span>
    <span>
      testing
       - 
      testing
    </span>
    <span
      class="badge light-blue badge-light-blue badge-pill mr-3"
      data-test="badge"
    >
      10
    </span>
  </li>,
  <li
    class="list-group-item d-flex justify-content-between list-group-item-undefined"
    data-test="list-group-item"
    data-testid="top-anagram-item"
    style="user-select: none;"
  >
    <span
      class="ml-3"
    >
      2
    </span>
    <span>
      tester
       - 
      tester
    </span>
    <span
      class="badge light-blue badge-light-blue badge-pill mr-3"
      data-test="badge"
    >
      9
    </span>
  </li>,
]
`;

exports[`Overall App Tests renders the app without data 1`] = `
<div>
  <div
    data-test="box"
  >
    <nav
      class="navbar navbar-expand-md"
      data-test="navbar"
      role="navigation"
    >
      <ul
        class="navbar-nav justify-content-around w-100"
        data-test="navbar-nav"
      >
        <li
          class="nav-item"
          data-test="nav-item"
        >
          <h2
            style="user-select: none;"
          >
            Community Anagrams
          </h2>
        </li>
      </ul>
    </nav>
    <div
      class="container"
      data-test="container"
    >
      <div
        class="card mt-5"
        data-test="card"
      >
        <div
          class="card-body"
          data-test="card-body"
        >
          <h4
            class="text-center mb-4"
          >
            Check Anagram
          </h4>
          <p
            class="text-center "
          >
            Check if two words are anagrams here.
             
            <strong>
              Do not include special characters and numbers.
            </strong>
          </p>
          <form>
            <div
              class="grey-text"
            >
              <div
                class="md-form"
              >
                
                <input
                  aria-disabled="false"
                  class="form-control"
                  data-test="input"
                  data-testid="first-word-input"
                  name="firstWord"
                  type="text"
                  value=""
                />
                <label
                  aria-labelledby=""
                  class=""
                  data-error=""
                  data-success=""
                  id=""
                >
                  First Word
                </label>
              </div>
              <div
                class="md-form"
              >
                
                <input
                  aria-disabled="false"
                  class="form-control"
                  data-test="input"
                  data-testid="second-word-input"
                  id="second-word-input"
                  name="secondWord"
                  type="text"
                  value=""
                />
                <label
                  aria-labelledby=""
                  class=""
                  data-error=""
                  data-success=""
                  for="second-word-input"
                  id=""
                >
                  Second Word
                </label>
              </div>
            </div>
            <div
              class="text-center"
            >
              <button
                class="btn-blue btn Ripple-parent"
                data-test="button"
                data-testid="anagram-submission-button"
                type="button"
              >
                Check
                <div
                  class="Ripple "
                  data-test="waves"
                  style="top: 0px; left: 0px; width: 0px; height: 0px;"
                />
              </button>
            </div>
          </form>
        </div>
      </div>
      <div
        class="mb-5"
        data-test="box"
      >
        <h5
          class="text-center pt-5 pb-5"
          data-testid="top-anagrams-board-header"
        >
          Top 10 Most Requested Anagram Checks
        </h5>
        <div
          class="h-100 w-100 d-flex justify-content-center align-middle"
          data-test="box"
        >
          <div
            class="spinner-border m-auto"
            role="status"
            style="width: 15rem; height: 15rem;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Overall App Tests should handle error when saving anagram 1`] = `
<div
  class="alert alert-danger text-center"
  data-testid="error-alert"
  role="alert"
/>
`;

exports[`Overall App Tests should handle not saving anagrams when invalid 1`] = `
<div
  class="alert alert-danger text-center"
  data-testid="error-alert"
  role="alert"
>
  Please remove any numbers or special symbols from your words and try again.
</div>
`;

exports[`Overall App Tests should handle saving anagrams when valid 1`] = `
<div
  class="alert alert-danger text-center"
  data-testid="result-alert"
  role="alert"
>
  Unfortunately 'racer' and 'carer' are not anagrams. Feel free to check other words and see if they are anagrams.
</div>
`;
